synapse/storage/databases/main/cache.py

```diff
@@ -271,6 +285,102 @@ def _invalidate_caches_for_event(
             self._attempt_to_invalidate_cache("get_thread_participated", (relates_to,))
             self._attempt_to_invalidate_cache("get_threads", (room_id,))
 
+    def _invalidate_caches_for_room_and_stream(
+        self, txn: LoggingTransaction, room_id: str
+    ) -> None:
+        """Invalidate caches associated with rooms, and stream to replication.
+
+        Used when we delete events in rooms, but don't know which events we've
+        deleted.
+        """
+
+        self._send_invalidation_to_replication(txn, DELETE_ROOM_CACHE_NAME, [room_id])
+        txn.call_after(self._invalidate_caches_for_room, room_id)
+
+    def _invalidate_caches_for_room(self, room_id: str) -> None:
+        """Invalidate caches associated with rooms.
+
+        Used when we delete events in rooms, but don't know which events we've
+        deleted.
+        """
+
+        self._invalidate_local_get_event_cache_all()  # type: ignore[attr-defined]
```

Comment:
Here and below: how did you find all these caches? I.e. how do we know that we've not missed any caches that hold events or event-derived data?

Reply:
(I guess the XXX comment above is meant to help keep this in sync with invalidations for a singular event, but for all I know `_invalidate_caches_for_event` might be missing something too. I guess it's a lot easier if they're both consistent though.)

Reply:
Mostly by looking for all caches with `room_id` in them. It would be good to try and make it more systematic, but I'm not really sure how.